### PR TITLE
Begin FreeBSD Support

### PR DESCRIPTION
**Pull request by time-killer-games**
Waiting for Josh to finish some stuff with JDI, in the meantime, you can make games for FreeBSD using this branch; I also finally got audio working but it requires building with clang instead of gcc. :D

jC26vln

Install instructions: https://enigma-dev.org/docs/Wiki/Install:FreeBSD

### CODE REVIEW

File: ENIGMAsystem/SHELL/Platforms/xlib/Makefile
Comment (fundies) [inline ENIGMAsystem/SHELL/Platforms/xlib/Makefile]: the pkg config stuff shouldnt need to be in the if here as its exactly same for both oses

File: shared/libpng-util/Makefile
Comment (fundies) [inline shared/libpng-util/Makefile]: you should only need this once in config.mk inside an if bsd. And *either* once in shell's makeffile also inside an if bsd *OR* in your ey file for bsd

File: shared/protos/Makefile
Comment (fundies) [inline shared/protos/Makefile]: no

File: shared/protos/Makefile
Comment (fundies) [inline shared/protos/Makefile]: no

File: CompilerSource/JDI/src/API/AST.cpp
Comment (fundies) [inline CompilerSource/JDI/src/API/AST.cpp]: if !defined(apple) && !defined(bsd)

File: CompilerSource/JDI/src/API/context.cpp
Comment (fundies) [inline CompilerSource/JDI/src/API/context.cpp]: there's likely a fix for this in upstream jdi so look at @JoshDreamland repo and see what hes changed

File: ENIGMAsystem/SHELL/Platforms/xlib/Linux/progdir.cpp
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/xlib/Linux/progdir.cpp]: Your gut was wrong. BSD can technically use readlink() but not only would the string argument be different but the OS does not enable the use of that function for this purpose by default, so I used something else that was more reliable and doesn't require any additional reconfiguring.

File: ENIGMAsystem/SHELL/Platforms/xlib/BSD/progdir.cpp
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/xlib/BSD/progdir.cpp]: Why are you always wrong? Jokes, with combined seriousness aside, I'm initializing the program_directory differently depending on whether it is Linux or BSD, because doing this on these platforms requires platform-specific code.

File: ENIGMAsystem/SHELL/Platforms/xlib/BSD/joystick.h
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/xlib/BSD/joystick.h]: Again, yes. You are just trolling. Save that for Discord and keep it off github where you are supposed to be actually helpful to the project.

File: ENIGMAsystem/SHELL/Platforms/xlib/Linux/joystick.h
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/xlib/Linux/joystick.h]: I can make this declared once if you want, but it doesn't belong in the general header unless Windows has it too. Because I am in a good mood I'll check and see if that's the case and will move it if it is. But really this problem existed before this pr between linux and windows, so it's not my job to do additional housekeeping in a pr that wasn't meant to address that.

File: CommandLine/libEGM/Makefile
Comment (time-killer-games) [reply CommandLine/libEGM/Makefile]: I must've done that by accident lel

File: CommandLine/emake/Makefile
Comment (fundies) [inline CommandLine/emake/Makefile]: you can use an or here

File: CommandLine/testing/Makefile
Comment (fundies) [inline CommandLine/testing/Makefile]: also or here

File: Config.mk
Comment (fundies) [inline Config.mk]: also or

File: ENIGMAsystem/SHELL/Makefile
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Makefile]: What?

**Comment by time-killer-games** [comment]
@fundies any ideas why running a game gives me this?
```
nomad@NomadBSD$ "/home/nomad/Key to Success 64-bit for Linux (Made with ENIGMA)/Key to Success"
ld-elf.so.1: /usr/local/lib/libalure.so.1: Undefined symbol "alcIsExtensionPresent"
```

**Comment by time-killer-games** [comment]
@fundies also @JoshDreamland said he has a few changes to make to JDI before its ready to be added to this pr. So we can't merge this until he does that and I have the correct JDI changes.

**Comment by time-killer-games** [comment]
@JoshDreamland is there a particular reason travis is claiming there is an undefined reference to `initialize_program_directory()` when I'm not getting that error locally on BSD or Linux, period?

**Comment by time-killer-games** [comment]
@JoshDreamland fundies threatened to get you to close this pr if i did this:

```
template<class RandomIt, class RandomFunc>
void random_shuffle(RandomIt first, RandomIt last, RandomFunc&& r)
{
    typename std::iterator_traits<RandomIt>::difference_type i, n;
    n = last - first;
    for (i = n-1; i > 0; --i) {
        using std::swap;
        swap(first[i], first[r(i+1)]);
    }
}
```

Instead of this:

#include <random>
#include <algorithm>
#include <iterator>
#include <iostream>
 
```
int main()
{
    std::vector<int> v = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
 
    std::random_device rd;
    std::mt19937 g(rd());
 
    std::shuffle(v.begin(), v.end(), g);
 
    std::copy(v.begin(), v.end(), std::ostream_iterator<int>(std::cout, " "));
    std::cout << "\n";
}
```
As a replacement to the deprecated random_shuffle as of c++14 and removed in c++17. I know how to do the former, i can just copy and paste it into a file, and the include it via a header and only build that source file when the makefiles are being used to build on FreeBSD. I don't know how to implement the equivalent by modifying the latter code example I copied from the cppreference website. Could you please show me a proper way to do this the way fundies prefers, so that he can finally stop harassing the god damned fuck out of me on discord? He is the most annoying pile of shit I swear to god. He won't leave me alone about this for fuck's sake.

**Comment by fundies** [comment]
I said nothing about closing your pr. I said i would tell him not to merge it because there's a stl function that you refuse to read docs on. Therefore, he simply shouldn't merge it until you do it correctly. Also, you begging josh for the answer isn't really learning but if he wants to spoon feed you, that's his prerogative

**Comment by time-killer-games** [comment]
Screenshot_2020-03-16_19-30-07
You know what you were implying. Stay away from my content and get out of my life.

**Comment by time-killer-games** [comment]
@JoshDreamland can you please remove fundies' comment?

File: ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp]: It's fine if you want to replace `std::random_shuffle` (deprecated) with `std::shuffle` but this needs to be connected with the user random seed. Review d6fa000a9c4323573703c888bc34e3b1e8780bdf and #1107 to understand why when this eventually fails the CI random test.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp]: @JoshDreamland told me to address that in a separate pr and he is in no rush for it to be done.

File: Unix.mk
Comment (fundies) [inline Unix.mk]: this should all be in config.mk under the same if else chain as the exts

File: shared/libpng-util/Makefile
Comment (fundies) [inline shared/libpng-util/Makefile]: shouldnt need this here. You might need to modify the shell makefile to include config.mk tho

File: ENIGMAsystem/SHELL/Universal_System/Extensions/IniFilesystem/Makefile
Comment (fundies) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/IniFilesystem/Makefile]: you dont need this include if you include config from shell's makefile

File: ENIGMAsystem/SHELL/Makefile
Comment (fundies) [inline ENIGMAsystem/SHELL/Makefile]: you shouldnt need these as theyre in config.mk

File: ENIGMAsystem/SHELL/Makefile
Comment (fundies) [inline ENIGMAsystem/SHELL/Makefile]: prolly good idea to move the mkdir setting to config.mk too

File: ENIGMAsystem/SHELL/Platforms/General/PFfilemanip.h
Comment (fundies) [inline ENIGMAsystem/SHELL/Platforms/General/PFfilemanip.h]: try to add comment why youre including new stuff

File: ENIGMAsystem/SHELL/Platforms/None/Makefile
Comment (fundies) [inline ENIGMAsystem/SHELL/Platforms/None/Makefile]: you shouldnt need this var as its already set in config.mk

File: shared/libpng-util/Makefile
Comment (fundies) [inline shared/libpng-util/Makefile]: you shouldnt need to explicity include this here as every root makefile should include config.mk before this file is ever included

Comment (time-killer-games) [reply shared/libpng-util/Makefile]: I actually already tried installing ENIGMA without including Config.mk in this Makefile and it errored with a reference to a missing "png.h" header. I need to include that so it can look for the headers and libs in the right place.

Comment (fundies) [reply shared/libpng-util/Makefile]: remove it and pastebin the build log

Comment (time-killer-games) [reply shared/libpng-util/Makefile]: Will do

Comment (time-killer-games) [reply shared/libpng-util/Makefile]: https://pastebin.com/rmrENVTw

Comment (time-killer-games) [reply shared/libpng-util/Makefile]: @JoshDreamland do you really care if I include Config.mk in this file? If so, I at least need to add to this Makefile the content from Config.mk that actually allows the headers and libs to be found. Are you ok with me doing either one of these things? BSD installs libs and includes to a different folder than Linux so I need the correct search paths.

**Comment by time-killer-games** [comment]
@JoshDreamland I still would like to address the stuff related to our random implementation in a follow up pr, and since you said you were fine with that, I did address everything else fundies wanted me to change, so this should be ready for a merge, unless you can find anything else that needs to be done first.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp
Comment (fundies) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp]: I see no need for this to be a template and is overall bad change because it generates a new seed every time you call it

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp:54]: i see no need for you to be a contributor to this project and is overall a bad change because you scared away every contributor we've ever had other than me and Robert. :P

File: ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp
Comment (fundies) [inline ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp:38]: as previously mentioned you can just assign to underlying data of the std::string instead of allocating your own buffer here

File: ENIGMAsystem/SHELL/Platforms/xlib/BSD/BSDjoystick.cpp
Comment (fundies) [inline ENIGMAsystem/SHELL/Platforms/xlib/BSD/BSDjoystick.cpp:47]: Try to avoid stubbing functions unless theyre required for compile and if they are please add DEBUG_MESSAGE informing users their game is using unimplemented functions

File: ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp:38]: I'm doing it correctly, so there's no reason to listen to you on this one. Marking it resolved. Have a good day. <3

Comment (fundies) [reply ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp:38]: don't resolve conversations unless both parties agree on solution

**Comment by fundies** [comment]
as previously mentioned you can just assign to underlying data of the std::string instead of allocating your own buffer here. I'm reposting this here because youre abusing gh's features to mark stuff as resolved that has not yet been resolved

**Comment by time-killer-games** [comment]
@fundies I'm blocking you because you used abusive behavior and harassment as a manipulative means to get me to unblock you to begin with.

**Comment by time-killer-games** [comment]
@JoshDreamland how do I achieve what fundies is asking so he stops harassing me? Please give a code example, because I've tried doing:

`char *buffer = path.data();`

Instead of:

`char *buffer = new char[length]();`

and without the delete[]. It prints an incorrect string with garbled characters.

File: ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp:38]: I'll do it as long as you state things I'm already aware of because you've said it 1000x on discord and only say it on here to carry over that same drama to a different website.

**Comment by time-killer-games** [comment]
@fundies you're abusing github in general as another outlet for you to be a public embarrassment. blocked without compensation. if you ask me to unblock you again, I'm reporting you to github staff,

File: ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp
Comment (fundies) [inline ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp:15]: should be able to just = {CTL_KERN, ...,. ...}; here

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/General/POSIX/BSD/BSDprogdir.cpp:15]: i find it funny how many times this little amount of code has been edited and improved. lel

**Review by RobertBColton — DISMISSED** [review]
I would like this pull request broken up into smaller requests because there's too many disparate changes going on here, as is the usual case of supporting any new platform with ENIGMA. I am also not completely convinced of the necessity in making so many specially adapted changes to support FreeBSD whose market share is what? Less than 0.01%?

**Comment by time-killer-games** [comment]
Thank you robert for the loving support. I knew I could count on you to cave into fundies and fuck everything up before @JoshDreamland was going to merge it.

**Comment by RobertBColton** [comment]
That's not how open source works, this isn't a race to get things into enigma-dev, all changes need thoroughly vetted. All I said is I want it in smaller pieces, because I am exhausted from having to backtrack all the time to patch over stuff that keeps breaking. If it was, most of it would probably be merged already.

**Comment by time-killer-games** [comment]
If you guys merged this, there would be nothing to complain about for either of us. The UNIX_BASED variable in the make files is a huge improvement and allows for much better platform checks. There is nothing about this pr that would need breaking up if it was merged. I'm perfectly happy to fix fundies and Josh's whining by attending to the random seed and whatnot but I'm not doing any of that until this is merged because working on this is giving us all a headache for no reason.

**Comment by time-killer-games** [comment]
@JoshDreamland since Robert didn't explain any specific ways to break this up, I have two questions.

1) Any chance you could list your objections to this pr so we may discuss ways to resolve them?

2) Are you going to enforce what Robert is telling me to do? All it seems to accomplish is create a lot of extra work for me needlessly, when there is enough to be done about this monster pr as it is.

3) There are actually 3 questions not 2, I can't count. If you do want me to break this up, could you explain to me more specifically what needs to be split up and how many pr's this should be turned into? That way I won't need to go back and make corrections on splitting it up how you guys wanted.

File: ENIGMAsystem/SHELL/Platforms/General/PFjoystick.h
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Platforms/General/PFjoystick.h:26]: I don't know if I agree with these being declared here. If you are going to do it anyway at least delete the old declarations of them wherever they were.

**Comment by RobertBColton** [comment]
Ideally I'd like to see. Feel free to combine the last two though and just send 3 pulls.
* Random stuff in one pull.
* Directory changes in another.
* Joysticks in a third one.
* Everything else in a fourth.

File: ENIGMAsystem/SHELL/Platforms/General/PFjoystick.h
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Platforms/General/PFjoystick.h:26]: If I were to go the extra mile and implement BSD joysticks, there is a good chance it will need initialization and handling code, and because OS X is based on BSD and both being Unix like as Linux is, we don't have any real way of knowing outside of an educated guess that Windows is *probably* the only platform which won't utilize these two functions, and that's completely ignoring the fact we are planning to do mobile as well which could potentially need these.

**Comment by time-killer-games** [comment]
@RobertBColton thanks now if @JoshDreamland could respond to me for the first time in months.

**Comment by time-killer-games** [comment]
@JoshDreamland any chance you could ignore Robert and fundies' rather lame excuses to put off merging this and just merge it? What has it been? 5 months? I've had way more patience on this than what I thought was possible for me.

**Comment by time-killer-games** [comment]
@JoshDreamland as more people are getting exposed to this project from the FreeBSD community, it would be rather nice if they could clone the most recent version of enigma and that won't be able to happen until you merge this pr which to be honest is passing still by some miracle and without any new conflicts so the reasons to delay merging this continue to get more thin.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp
Comment (JoshDreamland) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/DataStructures/data_structures.cpp:57]: Robert, please steward https://github.com/enigma-dev/enigma-dev/commit/d6fa000a9c4323573703c888bc34e3b1e8780bdf through to submission in another PR. I'm already not a fan of the huge amount of shuffling happening to support BSD; there's no sense blocking this change on fixing already-broken code rather than tweaking it slightly to work on more platforms in some capacity. I agree we should fix it—but not while we're moving earth to support BSD.

**Review by JoshDreamland — APPROVED** [review]
This is looking good. The UNIX_BASED Make variable in particular is a substantial improvement. I'm not a fan of the further Make-based platform specialization, but that's a marginal addition to a cleanup we're past due for, regardless. My concerns for this are minor. I think the broken(?) GCC configs will confuse people, and BSD oddisms are probably not a good default behavior when we can't identify the current platform (the general GNU versions are more likely to work without error).

File: Compilers/FreeBSD/gcc.ey
Comment (time-killer-games) [reply Compilers/FreeBSD/gcc.ey:1]: @JoshDreamland at the time, LateralGM was crashing because it needed those gcc.ey's to exist in order to run. Creating the files allowed it to open without crashing. However, we've made changes since that happened. I'll see if the changes effect whether this is still needed.

Comment (time-killer-games) [reply Compilers/FreeBSD/gcc.ey:1]: @JoshDreamland ok, i deleted  FreeBSD/gcc.ey and FreeBSD/gcc32.ey and no crash. Sounds like we must've fixed that since this was originally done.

File: CompilerSource/OS_Switchboard.h
Comment (time-killer-games) [reply CompilerSource/OS_Switchboard.h]: Is what i have now any better?

File: CommandLine/testing/Makefile
Comment (time-killer-games) [reply CommandLine/testing/Makefile:12]: @JoshDreamland considering FreeBSD uses all the same deps and API's as Linux, with the exception being the lack of real joystick/gamepad api, it should work I'd imagine. Though I'd like to know what I would need to do in order to test that theory, if it can be tested from the command line/emake or whatever. Never done it before.

Comment (time-killer-games) [reply CommandLine/testing/Makefile:12]: Clang doesn't need the OS_LIBS+=-lstdc++fs iirc.

**Comment by time-killer-games** [comment]
@JoshDreamland appveyor took another poop

File: Compilers/FreeBSD/gcc.ey
Comment (time-killer-games) [reply Compilers/FreeBSD/gcc.ey]: @JoshDreamland I stand corrected. I what i did differently last time i tried it, but i tried it again just now a clean install with no gcc.ey or gcc32.ey and it complained saying "Unsupported platform" but adding those two files back and running lgm again fixed that error. It seems lgm has it hardcoded the need for a gcc.ey file to both exist and be 100% readable. We've tried in the past renaming clang.ey to gcc.ey and that didn't work either.

**Comment by time-killer-games** [comment]
@JoshDreamland I've tested this pr in its current state and it is able to compile an empty 1 room game ok without errors. LGM didn't crash at all and everything seems to work. 

Although now we can have this platform more up to date and thanks to more recent changes we can also no longer require users to backup and delete their default make executable, which is another huge plus. 

In a follow up pr, I'm going to remove the need for creating google protocol buffer symlinks in the installation process of enigma for this platform, by updating the makefiles.

**Comment by time-killer-games** [comment]
@JoshDreamland appveyor keeps pooping

**Comment by time-killer-games** [comment]
@JoshDreamland are you able to understand what is wrong with appveyor? I am not at this point.

**Comment by JoshDreamland** [comment]
@RobertBColton @fundies are any of your requested changes unresolved?

I've restarted the AppVeyor.

**Comment by fundies** [comment]
@JoshDreamland I still have my concerns about his change to random/shuffle but I won't block it on that

**Pull request merged**
